cpp/ql/src/Likely Bugs/Underspecified Functions/ImplicitFunctionDeclaration.ql

```diff
@@ -0,0 +1,158 @@
+    parm.(PointerType).getBaseType().getUnspecifiedType())
+  or
+  pointerArgTypeMayBeUsed(arg.(ArrayType).getBaseType().getUnspecifiedType(),
+    parm.(PointerType).getBaseType().getUnspecifiedType())
+  or
+  // C11 arrays
+  pointerArgTypeMayBeUsed(arg.(PointerType).getBaseType().getUnspecifiedType(),
+    parm.(ArrayType).getBaseType().getUnspecifiedType())
+  or
+  pointerArgTypeMayBeUsed(arg.(ArrayType).getBaseType().getUnspecifiedType(),
+    parm.(ArrayType).getBaseType().getUnspecifiedType())
+}
+
+// This predicate holds whenever expression `arg` may be used to initialize
+// function parameter `parm` without need for run-time conversion.
+pragma[inline]
+predicate argMayBeUsed(Expr arg, ParameterDeclarationEntry pde) {
+  argTypeMayBeUsed(arg.getFullyConverted().getUnspecifiedType(), pde.getUnspecifiedType())
+}
+
+// True if this file (or header) was compiled as a C file
+predicate isCompiledAsC(File f) {
+  f.compiledAsC()
+  or
+  exists(File src | isCompiledAsC(src) | src.getAnIncludedFile() = f)
+}
+
+// True if function was ()-declared, but not (void)-declared or K&R-defined
+// or implicitly declared (i.e., lacking a prototype)
+predicate hasZeroParamDeclTooMany(Function f) {
+  exists(FunctionDeclarationEntry fde | fde = f.getADeclarationEntry() |
+    not fde.isImplicit() and
+    not fde.hasVoidParamList() and
+    fde.getNumberOfParameters() = 0 and
+    not fde.isDefinition()
+  )
+}
+
+// True if function was ()-declared, but not (void)-declared or K&R-defined
+predicate hasZeroParamDecl(Function f) {
+  exists(FunctionDeclarationEntry fde | fde = f.getADeclarationEntry() |
+    not fde.hasVoidParamList() and fde.getNumberOfParameters() = 0 and not fde.isDefinition()
+  )
+}
+
+bindingset[name]
+predicate notAlreadyReported(string name, FunctionCall fc) {
+  forall(FunctionDeclarationEntry fdeEx |
+    fdeEx.getName() = name and
+    not fdeEx.isImplicit()
+  |
+    // only report if not reported by cpp/futile-params
+    (
+      fdeEx.getNumberOfParameters() >= fc.getNumberOfArguments()
+      or
+      exists(Function f |
+        f = fc.getTarget() and
+        not exists(f.getBlock())
+        or
+        not hasZeroParamDeclTooMany(f)
+      )
+    ) and
+    // only report if not reported by cpp/too-few-arguments
+    (
+      fdeEx.getNumberOfParameters() <= fc.getNumberOfArguments()
+      or
+      exists(Function f |
+        f.isVarargs() or
+        f instanceof BuiltInFunction or
+        not hasZeroParamDecl(f)
+      )
+    ) and
+    // only report if not reported by cpp/mistyped-function-arguments
+    (
+      not hasZeroParamDecl(fc.getTarget())
+      or
+      forall(ParameterDeclarationEntry pde |
+        pde = fdeEx.getAParameterDeclarationEntry() and
+        pde.getIndex() < fc.getNumberOfArguments()
+      |
+        argMayBeUsed(fc.getArgument(pde.getIndex()), pde)
+      )
+    )
+  )
+}
+
+predicate sameLocation(Location loc1, Location loc2) {
```

Comment:
Expressions from macro expansions will have ambiguous location information, so we have to either exclude those expressions (use https://github.com/Semmle/ql/blob/183fd91a01c02ab12dcd9b25b97b14ccccadfafd/cpp/ql/src/semmle/code/cpp/commons/Exclusions.qll#L96) or find a better way to match a call to a FDE. What does the extractor team say?

Reply:
The idea of the `sameLocation` predicate came from the extractor team, but we did not specifically discuss how macros affected this - I will look into this!

Reply:
Ya, it seems like the only way to match the implicit FDE with the call that generated it is by using the location ... so I think the best solution for now is to exclude expressions generated from macro expansion